qldpc/codes.py

```diff
@@ -210,21 +218,139 @@ def dimension(self) -> int:
+
+        If passed a vector, compute the minimal Hamming distance between the vector and a code word.
+        Equivalently, we can interpret the given vector as an error, and find a minimal-weight
+        correction from decoding the syndrome induced by this vector.
+
+        Additional arguments, if applicable, are passed to a decoder.
+        """
+        if vector is not None:
+            # find the distance of the given vector from a code word
+            remainder = qldpc.decoder.decode(
+                self.matrix,
+                self.matrix @ self.field(vector),
+                **decoder_args,
+            )
+            return int(np.count_nonzero(remainder))
+
+        # effective syndrome: a trivial "actual" syndrome, and a nonzero overlap with a random word
+        effective_syndrome = np.zeros(self.num_checks + 1, dtype=int)
+        effective_syndrome[-1] = 1
+        _fix_decoder_args_for_nonprime_fields(decoder_args, self.field, bound_index=-1)
+
+        valid_candidate_found = False
+        while not valid_candidate_found:
+            # construct the effective check matrix enforcing
+            # (a) a trivial syndrome, and
+            # (b) a nonzero overlap with a random word
+            random_word = get_random_nontrivial_vec(self.field, self.num_bits)
+            effective_check_matrix = np.vstack([self.matrix, random_word]).view(np.ndarray)
+
+            # compute a low-weight candidate code word
+            candidate = qldpc.decoder.decode(
+                effective_check_matrix,
+                effective_syndrome,
+                **decoder_args,
+            )
+
+            # check whether we found a candidate candidate
+            actual_syndrome = effective_check_matrix @ candidate % self.field.order
+            valid_candidate_found = np.array_equal(actual_syndrome, effective_syndrome)
+
+        return int(np.count_nonzero(candidate))
 
-    def get_code_params(self) -> tuple[int, int, int]:
+    def get_code_params(
+        self, *, bound: int | None = None, **decoder_args: object
+    ) -> tuple[int, int, int, int]:
         """Compute the parameters of this code: [n,k,d].
 
         Here:
         - n is the number of data bits
         - k is the number of encoded ("logical") bits
         - d is the code distance
+
+        Keyword arguments are passed to the calculation of code distance.
         """
-        return self.num_bits, self.dimension, self.get_distance()
+        return (
+            self.num_bits,
+            self.dimension,
+            self.get_distance(bound=bound, vector=None, **decoder_args),
+            self.get_weight(),
+        )
+
+    def get_weight(self) -> int:
+        """Compute the weight of the largest check."""
+        row_max = max(np.count_nonzero(row) for row in self.matrix)
+        col_max = max(np.count_nonzero(col) for col in self.matrix.T)
+        return max(row_max, col_max)
```

Comment:
This is not exactly computing what it says in the docstring.  If all checks are weight 2 but one qubit is involved in 5 checks, this would return 5.  So really this is the maximum degree of the Tanner graph.  That's still a reasonable quantity to consider, but we should think about what exactly we want to return here, and what we want to call it.

Reply:
Yeah there is this confusion in the definition of LDPC codes. I think the classical definition counts both column and row sparsity but if you want the max weight of a check that is just row sparsity. Returning both gives the user flexibility.

Reply:
I might opt to go with the row-weight only.  If a user knows and cares about column weight that is easy enough to compute.

Reply:
I agree though that the definition of LDPC code involves both row and column sparsity.  But I am more worried about utility here than about definitions involving asymptotics.